### PR TITLE
Bug798475 - GNCAccountSel could have shortcuts

### PR DESCRIPTION
**Pull request by Bob-IT**
@christopherlam 
Chris, I think the first commit does what you asked for, can you give it a spin. I could not see a way of reusing the same code as the register but may have another look.

The second commit is what I suggested in your PR #1321 for using a secondary icon. I set the default for the GNCAccountSel to hide placeholders and hidden accounts but if the combo is set with such an account the settings are changed appropriately to show it.
Below is an image of what it looks like...
acc-sel
What do you think, worthwhile or not

### CODE REVIEW

**Comment by christopherlam** [comment]
:+1: :+1: :+1: much appreciated :)

**Comment by Bob-IT** [comment]
My last change has slowed down the loading so will need to rethink that and make changes later.
I did have another look at using account quickfill, with the use of the combo entry completion the only aspect that would be of use is using the shared list store but that would need to have a model filter on top so each GncAccountSel can be filtered while using a common list store. I will give it a try and report back.

File: gnucash/gnome-utils/gnc-account-sel.c [hunk 1]
```diff
@@ -144,33 +145,241 @@ combo_changed_cb (GNCAccountSel *gas, gpointer combo)
...
+    GNCAccountSel *gas = GNC_ACCOUNT_SEL(user_data);
+    GtkWidget *menu, *h_placeholder, *h_hidden;
+
+    if (icon_pos != GTK_ENTRY_ICON_SECONDARY)
+        return;
+
+    menu = gtk_menu_new ();
+    h_placeholder = gtk_check_menu_item_new_with_label (_("Hide Placeholder Accounts"));
+    gtk_check_menu_item_set_active (GTK_CHECK_MENU_ITEM(h_placeholder), gas->hide_placeholder);
+    h_hidden = gtk_check_menu_item_new_with_label (_("Hide Hidden Accounts"));
```
Comment (fellen) [inline gnucash/gnome-utils/gnc-account-sel.c:319]: Why not adding accellerators?
"Hide _Placeholder Accounts"
"Hide _Hidden Accounts"

File: gnucash/gnome-utils/gnc-account-sel.c [hunk 2]
```diff
@@ -144,33 +145,241 @@ combo_changed_cb (GNCAccountSel *gas, gpointer combo)
...
     gtk_combo_box_set_entry_text_column (GTK_COMBO_BOX(widget), ACCT_COL_NAME);
     g_signal_connect_swapped (gas->combo, "changed",
                               G_CALLBACK(combo_changed_cb), gas);
     gtk_container_add (GTK_CONTAINER(gas), widget);
 
+    entry = gtk_bin_get_child (GTK_BIN(gas->combo));
+    gtk_entry_set_icon_from_icon_name (GTK_ENTRY(entry), GTK_ENTRY_ICON_SECONDARY,
+                                       "preferences-system-symbolic");
+    gtk_entry_set_icon_tooltip_text (GTK_ENTRY(entry), GTK_ENTRY_ICON_SECONDARY,
+                                     _("Set visibility of Placeholder and Hidden Accounts."));
```
Comment (fellen) [inline gnucash/gnome-utils/gnc-account-sel.c:367]: "Set the visibility of placeholder and hidden accounts."
1. Article was missing.
2. No capitalization in tooltips

File: gnucash/gnome-utils/gnc-account-sel.c (hunk 1 above)
Comment (Bob-IT) [reply gnucash/gnome-utils/gnc-account-sel.c:319]: Did not think of it, added

File: gnucash/gnome-utils/gnc-account-sel.c (hunk 2 above)
Comment (Bob-IT) [reply gnucash/gnome-utils/gnc-account-sel.c:367]: Changed.

**Comment by Bob-IT** [comment]
I forgot to add that maybe with these changes it is too much for maint, possibly all but the last one, not really sure.
I was also thinking of renaming this widget to GncAccountSel which I think brings it in line with other widgets.

There is one other thing when using the account-quickfill list store, the appearance of the entries is tied to the preference setting GNC_PREF_SHOW_LEAF_ACCT_NAMES which is under 'Register Defaults->Other Defaults' so maybe that would need moving to be under 'Accounts->Labels'

File: gnucash/gnome-utils/gnc-account-sel.c [hunk 3]
```diff
@@ -136,151 +227,453 @@ gnc_account_sel_class_init (GNCAccountSelClass *klass)
...
+                valid = gtk_tree_model_iter_next (fmodel, &iter);
+            }
+            if (gas->prefix[0] == 0)
+                g_utf8_strncpy (gas->prefix, lower_entered_text, entered_len);
 
-    gas->initDone = TRUE;
+            g_free (lower_entered_text);
+        }
+        g_free (normal_entered_text);
+    }
```
Comment (jralls) [inline gnucash/gnome-utils/gnc-account-sel.c:411]: Everything between the previous comment and this one needs a refactor to be more intelligible.  The following is my attempt, please check to make sure that it's what you intended:
```
static char*
normalize_and_lower (char* utf8_string)
{
    char *normalized, *lowered;
    g_return_val_if_fail (utf8_string && *utf8_string, NULL);

    normalized = g_utf8_normalize (utf8_string, -1, G_NORMALIZE_ALL);
    g_free (utf8_string);
    if (!normalized)
        return NULL;
    lowered = g_utf8_strdown (normalized, -1);
    g_free (normalized);
    return lowered;
}

static inline gboolean
find_next_separator (const char* c, gunichar sep_char,
    size_t &offset)
{
    while (*c)
    {
        gboolean found = FALSE;
        gunichar uc = g_utf8_get_char (c);
        if (uc == sep_char)
        {
            found = TRUE;
            break;
        }
        c = g_utf8_next_char (c);
        ++(*offset);
    }
    return found;
}

/* Set gas->prefix from the account full name either to the next separator or
 * to the 
 * account_full_name is normalized and lowercased.
 */
static void
set_prefix_from_account_name (char* account_full_name, GNCAccountSel *gas,
                              int entered_len, gunichar sep_char, gint *prefix_len)
{
    gint item_offset_to_sep_char = entered_len;
    const char* c;
    gboolean found = FALSE;

    c = g_utf8_offset_to_pointer (account_full_name, item_offset_to_sep_char);
    item_offset_to_sep_char++;
    found = find_next_separator (c, sep_char, &item_offset_to_sep_char);

    if (!found)
        return;

    if (item_offset_to_sep_char < prefix_len)
    {
        *prefix_len = item_offset_to_sep_char;
        memset (gas->prefix, 0, BUFLEN);
        g_utf8_strncpy (gas->prefix, account_full_name, *prefix_len);
    }

    if (item_offset_to_sep_char == prefix_len)
    {
        char tmp_prefix[BUFLEN];

        memset (tmp_prefix, 0, BUFLEN);
        g_utf8_strncpy (tmp_prefix, account_full_name, prefix_len);

        while (g_strcmp0 (gas->prefix, tmp_prefix) != 0)
        {
            gchar *tmp = g_strdup (gas->prefix);
            prefix_len--;

            memset (tmp_prefix, 0, BUFLEN);
            g_utf8_strncpy (tmp_prefix, account_full_name, prefix_len);
            memset (gas->prefix, 0, BUFLEN);
            g_utf8_strncpy (gas->prefix, tmp, prefix_len);
            g_free (tmp);
        }
    }
}

static void
entry_insert_text_cb (GtkEntry *entry, const gchar *text, gint length,
                      gint *position, gpointer user_data)
 {
     GNCAccountSel *gas = GNC_ACCOUNT_SEL(user_data);
     GtkTreeModel *fmodel = gtk_combo_box_get_model (GTK_COMBO_BOX(gas->combo));
     const gchar *sep_char = gnc_get_account_separator_string ();
     gchar *entered_text;
     glong entered_len;
     gint prefix_len = G_MAXINT;
     gunichar sep_uchar = g_utf8_get_char (normalize_and_lower (sep_char));
     GtkTreeIter iter;
     gboolean valid;

     if (g_strcmp0 (text, sep_char) != 0)
         return;

     memset (gas->prefix, 0, BUFLEN);

     entered_text =
         normalize_and_lower (gtk_editable_get_chars (GTK_EDITABLE(entry), 0, -1));

     if (!entered_text)
         return;

     entered_len = g_utf8_strlen (entered_text, -1); //characters
     // Get the first item in the list
     valid = gtk_tree_model_get_iter_first (fmodel, &iter);

     // Walk through the list, reading each full name
     while (valid)
     {
         gchar *account_full_name, *lower_account_full_name;

         gtk_tree_model_get (fmodel, &iter, ACCT_COL_NAME, &account_full_name, -1);
         account_full_name = normalize_and_lower (account_full_name);

         if (!account_full_name)
             continue;

         if (g_str_has_prefix (account_full_name, entered_text))
             set_prefix_from_account_name (account_full_name, gas,
                                           entered_len, sep_char);

         g_free (account_full_name);
         valid = gtk_tree_model_iter_next (fmodel, &iter);
     }

     if (gas->prefix[0] == 0)
         g_utf8_strncpy (gas->prefix, entered_text, entered_len);

     g_free (entered_text);

     if (gas->prefix[0] != 0)
     {
         g_signal_handlers_block_by_func (GTK_EDITABLE(entry), (gpointer) entry_insert_text_cb, user_data);
         gtk_editable_delete_text (GTK_EDITABLE(entry), 0, -1);
         gtk_editable_set_position (GTK_EDITABLE(entry), 0);
         gtk_editable_insert_text (GTK_EDITABLE(entry), gas->prefix, -1, position);
         g_signal_handlers_unblock_by_func (GTK_EDITABLE(entry), (gpointer) entry_insert_text_cb, user_data);
         g_signal_stop_emission_by_name (GTK_EDITABLE(entry), "insert_text");
     }
 }
```

File: gnucash/gnome-utils/gnc-account-sel.c [hunk 4]
```diff
@@ -136,151 +227,453 @@ gnc_account_sel_class_init (GNCAccountSelClass *klass)
...
+        {
+            included = FALSE;
+        }
+    }
+    return included;
+}
 
-    g_signal_handlers_unblock_by_func (gas->combo, combo_changed_cb , gas);
+static gboolean
+visible_func (GtkTreeModel *model, GtkTreeIter *iter, gpointer user_data)
```
Comment (jralls) [inline gnucash/gnome-utils/gnc-account-sel.c:519]: Let's name this something a little more expressive like `account_is_visible_cb`.

File: gnucash/gnome-utils/gnc-account-sel.c [hunk 5]
```diff
@@ -136,151 +227,453 @@ gnc_account_sel_class_init (GNCAccountSelClass *klass)
...
+
+        if (gas->hide_placeholder && xaccAccountIsHidden (acc))
+            visible = FALSE;
+    }
+    return visible;
 }
 
 static void
-gas_filter_accounts (gpointer data, gpointer user_data)
+row_deleted_in_store_cb (GtkTreeModel *model, GtkTreePath *path, gpointer user_data)
```
Comment (jralls) [inline gnucash/gnome-utils/gnc-account-sel.c:541]: `row_deleted` sounds like a question. `maybe_delete_row_in_store_cb` or `ensure_row_deleted_in_store` would be better.

File: gnucash/gnome-utils/gnc-account-sel.c [hunk 6]
```diff
@@ -136,151 +227,453 @@ gnc_account_sel_class_init (GNCAccountSelClass *klass)
...
+}
 
-    if (atnd->gas->acctCommodityFilters)
+static void
+row_changed_in_store_cb (GtkTreeModel *model, GtkTreePath *path, GtkTreeIter *iter, gpointer user_data)
+{
+    GNCAccountSel *gas = GNC_ACCOUNT_SEL(user_data);
+
+    if (gas->selected_account_ref)
     {
```
Comment (jralls) [inline gnucash/gnome-utils/gnc-account-sel.c:571]: General comment applies pretty much everywhere: Deeply indented code is hard to read so if the function is done when the condition is false then reverse the condition and return. e.g.
```
    if (!gas->selected_account_ref)
        return;
```

File: gnucash/gnome-utils/gnc-account-sel.c [hunk 7]
```diff
@@ -387,7 +826,33 @@ gnc_account_sel_set_acct_filters (GNCAccountSel *gas, GList *typeFilters,
...
+                                         GList *excludeFilter)
+{
+    GtkTreeModel *fmodel;
+    GtkEntry *entry;
+
+    g_return_if_fail (gas != NULL);
+    g_return_if_fail (GNC_IS_ACCOUNT_SEL(gas));
+
+    fmodel = gtk_combo_box_get_model (GTK_COMBO_BOX(gas->combo));
+    entry = GTK_ENTRY(gtk_bin_get_child (GTK_BIN(gas->combo)));
```
Comment (jralls) [inline gnucash/gnome-utils/gnc-account-sel.c:844]: Neither of these variables is used.

File: gnucash/gnome-utils/gnc-account-sel.h [hunk 8]
```diff
@@ -38,24 +38,29 @@
 #define GNC_ACCOUNT_SEL_CLASS(klass)  G_TYPE_CHECK_CLASS_CAST (klass, GNC_TYPE_ACCOUNT_SEL, GNCAccountSelClass)
 #define GNC_IS_ACCOUNT_SEL(obj)       G_TYPE_CHECK_INSTANCE_TYPE (obj, GNC_TYPE_ACCOUNT_SEL)
 
+#define BUFLEN 1024
```
Comment (jralls) [inline gnucash/gnome-utils/gnc-account-sel.h:41]: Generic defines at global scope like this are a bad idea. Give it a name that's unlikely to be used elsewhere, perhaps `GNCACCTSEL_PREFIX_BUFLEN`.

Better still, put the struct in the implementation file and only expose the typedef in the header:
`typedef GncAccountSel struct _GncAccountSel`

**Review by jralls — CHANGES_REQUESTED** [review]

**Comment by Bob-IT** [comment]
> @Bob-IT Ping!

Been busy but will try and have a look this week, probably move this to master.

File: gnucash/gnome-utils/gnc-account-sel.c (hunk 5 above)
Comment (Bob-IT) [reply gnucash/gnome-utils/gnc-account-sel.c:541]: Changed.

File: gnucash/gnome-utils/gnc-account-sel.c (hunk 6 above)
Comment (Bob-IT) [reply gnucash/gnome-utils/gnc-account-sel.c:571]: OK.

File: gnucash/gnome-utils/gnc-account-sel.h (hunk 8 above)
Comment (Bob-IT) [reply gnucash/gnome-utils/gnc-account-sel.h:41]: OK, I thought I had done this in one of my changes locally, done now.

File: gnucash/gnome-utils/gnc-account-sel.c (hunk 3 above)
Comment (Bob-IT) [reply gnucash/gnome-utils/gnc-account-sel.c:411]: I see what you mean, I have a tendency to just add parts when I come across I need them. Will try and do more refactoring when I push.

File: gnucash/gnome-utils/gnc-account-sel.c [hunk 9]
```diff
@@ -144,10 +144,172 @@ combo_changed_cb (GNCAccountSel *gas, gpointer combo)
...
+
+    entered_text = gtk_editable_get_chars (GTK_EDITABLE(entry), 0, -1);
+
+    if (entered_text && *entered_text)
+    {
+        gchar *normal_entered_text = g_utf8_normalize (entered_text, -1,
+                                                       G_NORMALIZE_ALL);
+        if (normal_entered_text)
+        {
+            gchar *lower_entered_text = g_utf8_strdown (normal_entered_text, -1);
```
[the remark this answers is not in the document]
Comment (Bob-IT) [reply gnucash/gnome-utils/gnc-account-sel.c:320]: Changed.

File: gnucash/gnome-utils/gnc-account-sel.c [hunk 10]
```diff
@@ -144,10 +144,172 @@ combo_changed_cb (GNCAccountSel *gas, gpointer combo)
...
+            glong entered_len = g_utf8_strlen (lower_entered_text, -1); //characters
+            gunichar sep_char = gnc_get_account_separator ();
```
[the remark this answers is not in the document]
Comment (Bob-IT) [reply gnucash/gnome-utils/gnc-account-sel.c:322]: Not sure about this, I have used both forms of getting the separator character, at the top I used `const gchar *sep_char = gnc_get_account_separator_string ();` and later on used `sep_unichar = gnc_get_account_separator ();`

**Comment by Bob-IT** [comment]
@jralls Thanks for looking at this, I think I have made the changes required, just pondering the the gunichar comment.
I have left this on maint and if OK can push after the next release.

**Comment by jralls** [comment]
Merged into master.

**Pull request closed without merging**